File: 03 Animal/csharp/Branch.cs
namespace Animal
{
    public class Branch
    {
        public string Text { get; set; }

        public bool IsEnd => Yes == null && No == null;

        public Branch Yes { get; set; }

        public Branch No { get; set; }

        public override string ToString()
        {
            return $"{Text} : IsEnd {IsEnd}";
        }
    }
}

File: 03_Animal/vbnet/Animal/Branch.vb
Public Class Branch
    Public Property Text As String

    Public ReadOnly Property IsEnd As Boolean
        Get
            Return Yes Is Nothing AndAlso No Is Nothing
        End Get
    End Property

    Public Property Yes As Branch
    Public Property No As Branch

    ' Allows walking all the descendants recursively
    Public Iterator Function DescendantTexts() As IEnumerable(Of String)
        If Yes IsNot Nothing Then
            Yield Yes.Text
            For Each childText In Yes.DescendantTexts
                Yield childText
            Next
        End If

        If No IsNot Nothing Then
            Yield No.Text
            For Each childText In No.DescendantTexts
                Yield childText
            Next
        End If
    End Function
End Class
